File: mynlp-segment/src/main/java/com/mayabot/nlp/segment/crf/FeatureTemplate.java
/*
 * Copyright 2018 mayabot.com authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mayabot.nlp.segment.crf;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * 特征模板
 *
 * @author hankcs
 */
public class FeatureTemplate implements Serializable {

    public static int serialVersionUID = 1;

    /**
     * 用来解析模板的正则表达式
     */
    static final Pattern pattern = Pattern.compile("%x\\[(-?\\d*),(\\d*)]");


    String template;
    /**
     * 每个部分%x[-2,0]的位移，其中int[0]储存第一个数（-2），int[1]储存第二个数（0）
     */
    ArrayList<int[]> offsetList;

    List<String> delimiterList;

    public FeatureTemplate() {
    }

    public static FeatureTemplate create(String template) {
        FeatureTemplate featureTemplate = new FeatureTemplate();
        featureTemplate.delimiterList = new LinkedList<String>();
        featureTemplate.offsetList = new ArrayList<int[]>(3);
        featureTemplate.template = template;
        Matcher matcher = pattern.matcher(template);
        int start = 0;
        while (matcher.find()) {
            featureTemplate.delimiterList.add(template.substring(start, matcher.start()));
            start = matcher.end();
            featureTemplate.offsetList.add(new int[]{Integer.parseInt(matcher.group(1)), Integer.parseInt(matcher.group(2))});
        }
        return featureTemplate;
    }

    public char[] generateParameter(Table table, int current) {
        StringBuilder sb = new StringBuilder();
        int i = 0;
        for (String d : delimiterList) {
            sb.append(d);
            int[] offset = offsetList.get(i++);
            sb.append(table.get(current + offset[0], offset[1]));
        }

        char[] o = new char[sb.length()];
        sb.getChars(0, sb.length(), o, 0);

        return o;
    }


    @Override
    public String toString() {
        String sb = "FeatureTemplate{" + "template='" + template + '\'' +
                ", delimiterList=" + delimiterList +
                '}';
        return sb;
    }
}

File: mynlp/src/main/java/com/mayabot/nlp/segment/lexer/crf/FeatureTemplate.kt
/*
 * Copyright 2018 mayabot.com authors. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mayabot.nlp.segment.lexer.crf

import java.util.regex.Pattern

/**
 * @author jimichan
 */
class FeatureTemplateGroup(templates: List<String>) {

    val size: Int

    val list = ArrayList<FeatureTemplate>()

    init {
        for (template in templates) {
            list.add(FeatureTemplate(template))
        }
        this.size = templates.size
    }

    companion object {
        val BOS = arrayOf("_B-1", "_B-2", "_B-3", "_B-4", "_B-5", "_B-6", "_B-7", "_B-8")
        val EOS = arrayOf("_B+1", "_B+2", "_B+3", "_B+4", "_B+5", "_B+6", "_B+7", "_B+8")
    }

}


// U6:%x[1,0]/%x[2,0]
/**
 * @author jimichan
 */
class FeatureTemplate(template: String) {

    val list = ArrayList<FeatureTemplateElement>(10)

    private val pattern = Pattern.compile("%x\\[(-?\\d*),(\\d*)]")

    init {
        val matcher = pattern.matcher(template)
        var start = 0
        while (matcher.find()) {
            val offset = matcher.start()
            val end = matcher.end()
            if (offset > start) {
                list.add(FeatureTemplateElement(template.substring(start, offset)))
            }

            list.add(FeatureTemplateElement(matcher.group(1).toInt(), matcher.group(2).toInt()))

            start = end
        }

        val first = list.removeAt(0)

        list.add(first)
    }

    override fun toString(): String {
        return list.joinToString(separator = "")
    }
}

enum class FeatureTemplateElementType {
    String, Offset
}

class FeatureTemplateElement {

    constructor(value: String) {
        if (value.startsWith("U") && value.endsWith(":")) {
            this.value = ":" + value.substring(0, value.length - 1)
        } else {
            this.value = value
        }

    }

    constructor(offset: Int, col: Int) {
        this.offset = offset
        this.col = col
        this.type = FeatureTemplateElementType.Offset
    }

    var value: String = ""
    var offset: Int = 0
    var col: Int = 0
    var type = FeatureTemplateElementType.String

    override fun toString(): String {
        if (type == FeatureTemplateElementType.String) {
            return value
        } else {
            return "%x[$offset,$col]"
        }
    }
}
